### ISSUE TITLE
RFC 3.1: Decouple cookies from a controller

### ISSUE THREAD

**Issue by robertpustulka** [cakephp/cakephp#6717]
Following https://github.com/cakephp/cakephp/pull/6712#issuecomment-107870108

I was thinking about moving all cookie logic to a single handler. Currently we have cookies logic spread across CookieComponent, Request and Response.

Decoupling advanced cookies logic from a component to an independent handler would allow us to use cookies encryption etc outside controller layer. This "CookieHandler" could be attached to request the same way as sessions are.

The only problem is how to keep BC with 3.0, where cookies API is slightly different in Request and Response. This obviously should be simplified... I'm wondering how to do it the right way.

**Comment by markstory** [comment]
Cookies have to be in the Request and Response classes. As you need to read cookies that came in on the request and send cookies in the response.

The CookieComponent exists for a few reasons:

* Add encrypted cookies.
* Add array access into cookie data.
* Backwards compatibility with older versions of CakePHP.

How would you see a cookie handler working?

**Comment by robertpustulka** [comment]
@markstory 
> How would you see a cookie handler working?

All the `Request::cookie()` and `CookieComponent` functionality in one place. Preferably accesible via Request/Response. So I could for example do complex things with cookies (in view Cells, custom event listeners etc) not having to worry about loading `CookieComponent` and/or Controller dependency. 
There has been done something similar with sessions and probably wil be done (or have been?) with auth logic AFAIK.

**Comment by markstory** [comment]
@robertpustulka Ok, I think we'd have two different cookie 'jars' as the features you need around cookies for in the request are very different than those in the response. CookieComponent can work around this by having access to both the request & response.

I would want to avoid a situation where the following was possible:

```php
$this->request->cookies()->set('key', 'value');
```

This is nonsensical but should not be possible.

**Comment by robertpustulka** [comment]
Yeah, I thought about these request cookies.
So maybe split this functionality into some `CookieAccessor` for getting cookies from a request and `CookieMutator` to manipulate cookies to be sent with a response. But should CookieComponent be able  to manipulate request cookies? Does it make sense?

**Comment by markstory** [comment]
I would think that CookieComponent would use the request + response to read/write cookies with the existing API it provides.

**Comment by robertpustulka** [comment]
Currently we can have multiple CookieComponents configured and used to manipulate different cookies with various settings. I'm wondering how would it be possible using single cookies handler attached to Request/Response objects.

First thing that came to my mind was splitting this functionality into two elements. `CookieJar` (or two of them - one with read-only access for Request, another for Response) would keep and provide dot-notation access to cookies with the ability to set cookie options (path, expires, etc) explicitly. Another class - let's call it `CookieHandler` - would encapsulate some default options for cookies and use `CookieJar` to manipulate properly pre-configured cookies.

Eventually multiple `CookieHandler` objects could co-exist in an application. If there was a need for some more complex cookie handling, a client could spawn its own `CookieHandler` and access/store cookies without the need to depend on Controller/Component layer. To keep BC with older CakePHP versions a `CookieComponent` could create its own `CookieHandler` and proxy its API to the handler.

In short version - wrap cookies array into some object and move CookieComponent functionality into Controller independent class.

I'll try to make some code samples later.

**Comment by robertpustulka** [comment]
Another thing came to my mind...

CookieComponent's role is just to grab some data, pack it into some string form with encryption, set a path/domain/expires, name it and send to response. That makes it just a normal cookie. So instead of making CookieHandlers we could just have regular cookie objects. Each cookie object would be responsible for manipulating its own properties (value, expires etc). `CookieComponent` would be a wrapper for a `Cookie` object and keep its API. 

And there would be two CookieJars - one for a Request with read-only access another for Response.
In other words - Cookie objects would be our "cookie entities", Jars would become "cookie repositories".

**Comment by robertpustulka** [comment]
There's a PR open, I'm closing this then.
